crates/wasmparser/src/validator/operators.rs

```diff
@@ -23,75 +23,184 @@
 // the various methods here.
 
 use crate::{
-    limits::MAX_WASM_FUNCTION_LOCALS, BinaryReaderError, BlockType, MemoryImmediate, Operator,
-    Result, SIMDLaneIndex, ValType, WasmFeatures, WasmFuncType, WasmModuleResources,
+    limits::MAX_WASM_FUNCTION_LOCALS, BinaryReaderError, BlockType, BrTable, Ieee32, Ieee64,
+    MemoryImmediate, Operator, Result, SIMDLaneIndex, ValType, VisitOperator, WasmFeatures,
+    WasmFuncType, WasmModuleResources, V128,
 };
 
-/// A wrapper around a `BinaryReaderError` where the inner error's offset is a
-/// temporary placeholder value. This can be converted into a proper
-/// `BinaryReaderError` via the `set_offset` method, which replaces the
-/// placeholder offset with an actual offset.
-pub(crate) struct OperatorValidatorError(pub(crate) BinaryReaderError);
-
 /// Create an `OperatorValidatorError` with a format string.
 macro_rules! format_op_err {
-    ( $( $arg:expr ),* $(,)* ) => {
-        OperatorValidatorError::new(format!( $( $arg ),* ))
+    ( $offset:expr, $( $arg:expr ),* $(,)* ) => {
+        BinaryReaderError::new(format!( $( $arg ),* ), $offset)
     }
 }
 
 /// Early return an `Err(OperatorValidatorError)` with a format string.
 macro_rules! bail_op_err {
-    ( $( $arg:expr ),* $(,)* ) => {
-        return Err(format_op_err!( $( $arg ),* ))
+    ( $offset:expr, $( $arg:expr ),* $(,)* ) => {
+        return Err(format_op_err!( $offset, $( $arg ),* ))
     }
 }
 
-impl OperatorValidatorError {
-    /// Create a new `OperatorValidatorError` with a placeholder offset.
-    pub(crate) fn new(message: impl Into<String>) -> Self {
-        let offset = std::usize::MAX;
-        let e = BinaryReaderError::new(message, offset);
-        OperatorValidatorError(e)
+type OperatorValidatorResult<T> = std::result::Result<T, BinaryReaderError>;
+
+/// The Wasm control list and type stack.
+pub(crate) struct Stack {
+    /// This is a list of flags for wasm features which are used to gate various
+    /// instructions.
+    features: WasmFeatures,
+    /// The `control` list is the list of blocks that we're currently in.
+    control: Vec<Frame>,
+    /// The `operands` is the current type stack.
+    operands: Vec<Option<ValType>>,
+}
```

Comment:
How come this type was split out from the operators validator?

Reply:
It was to resolve some borrow checking issues. Also I was under the impression that it cleaned up the code a bit actually. I often see myself splitting out grouped functionality into their own types to resolve borrow checking issues elegantly.